File: backend/src/controllers/auth.controller.ts
import type { Response } from "express";
import type {
  AdminRegisterValidatedRequest,
  LoginValidatedRequest,
  RefreshTokenValidatedRequest,
  RegisterValidatedRequest,
} from "../middleware/auth.validation.js";
import type { AuthenticatedRequest } from "../middleware/auth.middleware.js";
import {
  clearGuestSessionCookie,
  type GuestSessionRequest,
} from "../middleware/guest-session.js";
import { AppError } from "../errors/app-error.js";
import {
  getCurrentAuthUser,
  loginAdmin,
  loginUser,
  logoutSession,
  mergeGuestCartForUser,
  refreshAccessToken,
  registerAdmin,
  registerUser,
} from "../services/auth.service.js";
import type {
  AuthUserResponse,
  LoginResponseData,
  LogoutResponseData,
  RefreshTokenResponseData,
  RegisterResponseData,
} from "../types/auth.types.js";
import type { ErrorResponse } from "../types/product.types.js";

export async function register(
  req: RegisterValidatedRequest & GuestSessionRequest,
  res: Response<RegisterResponseData | ErrorResponse>,
): Promise<void> {
  const validatedBody = req.validatedBody;

  if (validatedBody === undefined) {
    throw AppError.badRequest("Invalid register payload");
  }

  const responseData = await registerUser(validatedBody);

  try {
    if ((await mergeGuestCartForUser(responseData.user.id, req.guestId)) === true) {
      clearGuestSessionCookie(res);
    }
  } catch (error) {
    console.warn("Failed to merge guest cart after register:", error);
  }

  res.status(201).json(responseData);
}

export async function login(
  req: LoginValidatedRequest & GuestSessionRequest,
  res: Response<LoginResponseData | ErrorResponse>,
): Promise<void> {
  const validatedBody = req.validatedBody;

  if (validatedBody === undefined) {
    throw AppError.badRequest("Invalid login payload");
  }

  const responseData = await loginUser(validatedBody);

  try {
    if ((await mergeGuestCartForUser(responseData.user.id, req.guestId)) === true) {
      clearGuestSessionCookie(res);
    }
  } catch (error) {
    console.warn("Failed to merge guest cart after login:", error);
  }

  res.status(200).json(responseData);
}

export async function adminRegister(
  req: AdminRegisterValidatedRequest & GuestSessionRequest,
  res: Response<RegisterResponseData | ErrorResponse>,
): Promise<void> {
  const validatedBody = req.validatedBody;

  if (validatedBody === undefined) {
    throw AppError.badRequest("Invalid admin register payload");
  }

  const responseData = await registerAdmin(validatedBody);

  try {
    if ((await mergeGuestCartForUser(responseData.user.id, req.guestId)) === true) {
      clearGuestSessionCookie(res);
    }
  } catch (error) {
    console.warn("Failed to merge guest cart after admin register:", error);
  }

  res.status(201).json(responseData);
}

export async function adminLogin(
  req: LoginValidatedRequest & GuestSessionRequest,
  res: Response<LoginResponseData | ErrorResponse>,
): Promise<void> {
  const validatedBody = req.validatedBody;

  if (validatedBody === undefined) {
    throw AppError.badRequest("Invalid admin login payload");
  }

  const responseData = await loginAdmin(validatedBody);

  try {
    if ((await mergeGuestCartForUser(responseData.user.id, req.guestId)) === true) {
      clearGuestSessionCookie(res);
    }
  } catch (error) {
    console.warn("Failed to merge guest cart after admin login:", error);
  }

  res.status(200).json(responseData);
}

export async function refreshToken(
  req: RefreshTokenValidatedRequest,
  res: Response<RefreshTokenResponseData | ErrorResponse>,
): Promise<void> {
  const validatedBody = req.validatedBody;

  if (validatedBody === undefined) {
    throw AppError.badRequest("Invalid refresh token payload");
  }

  const responseData = await refreshAccessToken(validatedBody);
  res.status(200).json(responseData);
}

export async function logout(
  req: RefreshTokenValidatedRequest,
  res: Response<LogoutResponseData | ErrorResponse>,
): Promise<void> {
  const validatedBody = req.validatedBody;

  if (validatedBody === undefined) {
    throw AppError.badRequest("Invalid logout payload");
  }

  const responseData = await logoutSession(validatedBody);
  res.status(200).json(responseData);
}

export async function me(
  req: AuthenticatedRequest,
  res: Response<AuthUserResponse | null | ErrorResponse>,
): Promise<void> {
  const userId = req.authUserId;

  if (userId === undefined) {
    res.status(200).json(null);
    return;
  }

  const responseData = await getCurrentAuthUser(userId);
  res.status(200).json(responseData);
}

File: backend/src/main/java/com/season/app/controller/AuthController.java
package com.season.app.controller;

import com.season.app.dto.auth.*;
import com.season.app.service.AuthService;
import com.season.app.service.CartService;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.security.Principal;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    private AuthService authService;

    @Autowired
    private CartService cartService;

    private void clearGuestCookie(HttpServletResponse response) {
        Cookie cookie = new Cookie("guestId", null);
        cookie.setHttpOnly(true);
        cookie.setPath("/");
        cookie.setMaxAge(0);
        response.addCookie(cookie);
    }

    @PostMapping("/login")
    public ResponseEntity<LoginResponseData> login(
            @Valid @RequestBody LoginInput input,
            @CookieValue(name = "guestId", required = false) String guestId,
            HttpServletResponse response) {
            
        LoginResponseData data = authService.loginUser(input);
        if (cartService.mergeGuestCartIntoUserCart(data.getUser().getId(), guestId)) {
            clearGuestCookie(response);
        }
        return ResponseEntity.ok(data);
    }

    @PostMapping("/register")
    public ResponseEntity<RegisterResponseData> register(
            @Valid @RequestBody RegisterInput input,
            @CookieValue(name = "guestId", required = false) String guestId,
            HttpServletResponse response) {
            
        RegisterResponseData data = authService.registerUser(input);
        if (cartService.mergeGuestCartIntoUserCart(data.getUser().getId(), guestId)) {
            clearGuestCookie(response);
        }
        return ResponseEntity.status(HttpStatus.CREATED).body(data);
    }

    @PostMapping("/admin/register")
    public ResponseEntity<RegisterResponseData> adminRegister(
            @Valid @RequestBody AdminRegisterInput input,
            @CookieValue(name = "guestId", required = false) String guestId,
            HttpServletResponse response) {
            
        RegisterResponseData data = authService.registerAdmin(input, input.getAdminSecret());
        if (cartService.mergeGuestCartIntoUserCart(data.getUser().getId(), guestId)) {
            clearGuestCookie(response);
        }
        return ResponseEntity.status(HttpStatus.CREATED).body(data);
    }

    @PostMapping("/admin/login")
    public ResponseEntity<LoginResponseData> adminLogin(@Valid @RequestBody LoginInput input) {
        LoginResponseData response = authService.loginUser(input);
        if (!"admin".equals(response.getUser().getRole())) {
            throw new AuthService.AuthServiceError("Admin account is required", HttpStatus.FORBIDDEN.value());
        }
        return ResponseEntity.ok(response);
    }

    @PostMapping("/refresh-token")
    public ResponseEntity<RefreshTokenResponseData> refreshToken(@Valid @RequestBody RefreshTokenInput input) {
        RefreshTokenResponseData response = authService.refreshAccessToken(input);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/logout")
    public ResponseEntity<LogoutResponseData> logout(@Valid @RequestBody RefreshTokenInput input) {
        LogoutResponseData response = authService.logoutSession(input);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/me")
    public ResponseEntity<AuthUserResponse> me(Principal principal) {
        if (principal == null || principal.getName() == null) {
            return ResponseEntity.ok(null);
        }
        AuthUserResponse response = authService.getCurrentAuthUser(principal.getName());
        return ResponseEntity.ok(response);
    }
}